### PR TITLE
enhancement(es sink): separate aws support in es & prometheus sink

### PR DESCRIPTION
**Pull request by suikammd**
Currently, the sinks-elasticsearch feature in Vector enables AWS support by default, which may not be necessary for all users. This leads to an increase in compiled time and compiled code size for users who only require the normal Elasticsearch functionality.
So, I make make sinks-elasticsearch feature disable aws on default and enable aws when aws-core feature enabled. Same for sinks-prometheus.

### CODE REVIEW

**Comment by jszwedko** [comment]
Hi @suikammd !

We chatted about this PR a bit today. We are open to this contribution moving the AWS bits of these sinks behind the `aws-core` feature flag, but we'd like to do it in a way that doesn't require a lot of feature flagging within the sink implementations as this makes maintenance more difficult. Ideally this would be implemented in a way where we only need the feature flag present in 1 or 2 places per sink (your changes to the ES sink, for example introduce the flag in 36 places). This will require some significant refactoring to isolate the bits that need to be flagged though. What do you think?

File: src/sinks/elasticsearch/common.rs [hunk 1]
```diff
@@ -1,6 +1,8 @@
 use std::collections::HashMap;
 
+#[cfg(feature = "aws-core")]
 use aws_credential_types::provider::SharedCredentialsProvider;
```
Comment (neuronull) [inline src/sinks/elasticsearch/common.rs:4]: **❓ question:** ‏ Is this still needed? If it is, the feature flagged use statement could conceivably be removed from the top of the file here and the type or trait path be inlined where needed, or if in a block, the use statement can be moved there.

```suggestion
```

File: src/sinks/elasticsearch/config.rs [hunk 2]
```diff
@@ -6,8 +6,9 @@ use std::{
 use futures::{FutureExt, TryFutureExt};
 use vector_config::configurable_component;
 
+#[cfg(feature = "aws-core")]
+use crate::aws::RegionOrEndpoint;
```
Comment (neuronull) [inline src/sinks/elasticsearch/config.rs:10]: **💬 suggestion:** ‏similarly here, I think we can remove the `use` statement and just directly reference the `crate::aws::RegionOrEndpoint` in the already feature gated code in the file.

File: src/sinks/elasticsearch/mod.rs [hunk 3]
```diff
@@ -24,6 +24,7 @@ use snafu::Snafu;
 use vector_common::sensitive_string::SensitiveString;
 use vector_config::configurable_component;
 
+#[cfg(feature = "aws-core")]
 use crate::aws::AwsAuthentication;
```
Comment (neuronull) [inline src/sinks/elasticsearch/mod.rs:28]: **💬 suggestion:** ‏same

File: src/sinks/prometheus/mod.rs [hunk 4]
```diff
@@ -8,6 +8,7 @@ pub(crate) mod remote_write;
 
 use vector_config::configurable_component;
 
+#[cfg(feature = "aws-core")]
 use crate::aws::AwsAuthentication;
```
Comment (neuronull) [inline src/sinks/prometheus/mod.rs:12]: same

File: src/sinks/util/auth.rs [hunk 5]
```diff
@@ -0,0 +1,14 @@
+#[cfg(feature = "aws-core")]
+use aws_credential_types::provider::SharedCredentialsProvider;
+#[cfg(feature = "aws-core")]
+use aws_types::region::Region;
```
Comment (neuronull) [inline src/sinks/util/auth.rs:4]: **💬 suggestion:** ‏similarly, I think this should work

```suggestion
```

File: src/sinks/util/auth.rs [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+
+#[derive(Debug, Clone)]
+pub enum Auth {
+    Basic(crate::http::Auth),
+    #[cfg(feature = "aws-core")]
+    Aws {
+        credentials_provider: SharedCredentialsProvider,
+        region: Region,
+    },
```
Comment (neuronull) [inline src/sinks/util/auth.rs:13]:
```suggestion
    Aws {
        credentials_provider: aws_credential_types::provider::SharedCredentialsProvider,
        region: aws_types::region::Region,
    },
```

File: src/sinks/prometheus/remote_write.rs [hunk 7]
```diff
@@ -14,16 +14,18 @@ use vector_config::configurable_component;
 use vector_core::{ByteSizeOf, EstimatedJsonEncodedSizeOf};
 
 use super::collector::{self, MetricCollector as _};
+#[cfg(feature = "aws-core")]
+use crate::aws::RegionOrEndpoint;
```
Comment (neuronull) [inline src/sinks/prometheus/remote_write.rs:18]: Same

**Review by neuronull — COMMENTED** [review]
Thanks for looking into reducing the amount of feature flag gates added. I left some suggestions on how to hopefully reduce further.

File: src/sinks/elasticsearch/common.rs (hunk 1 above)
Comment (suikammd) [reply src/sinks/elasticsearch/common.rs]: hi @neuronull thx for you advice. I modified all you mentioned code.

**Comment by neuronull** [comment]
Thanks, looking better. There are still two suggestions I had to `util/auth.rs` .

Also, if in the future you could refrain from force-pushing, that would be greatly appreciated. It makes reviewing incremental changes much easier.

**Comment by suikammd** [comment]
Sorry for missing some code, there should be no left now :)

**Review by jszwedko — APPROVED** [review]
Thanks for the additional refactoring @suikammd !

**Comment by jszwedko** [comment]
@suikammd small note: please try to refrain from force pushing 😄  It makes reviewing more difficult since reviewers can't just focus on new changes.

**Comment by suikammd** [comment]
@jszwedko Oh, sorry for that, I did a habitually rebase from master, then I add `format code` commit. Maybe it's enough to just review this commit.

**Comment by jszwedko** [comment]
Thanks @suikammd !

**Pull request merged**